lib/js/protocol/MessageFrameAssembler.js

```diff
@@ -30,55 +30,154 @@
     }
 
     /**
-    * @param {SdlPacket} sdlPacket
-    */
-    handleFirstDataFrame(sdlPacket){
+     * Handles the first data frame. Sets the expected frame count and data length.
+     * @param {SdlPacket} sdlPacket 
+     * @private
+     */
+    _handleFirstDataFrame(sdlPacket) {
+
+        /** @type {Uint8Array} */
+        let payload = sdlPacket.getPayload();
+
+        //first 4 8-bit integers contain the data length.
+        let dataLength = (payload[0] & 0x0F) << 24; //8-bit x 3
+        dataLength += (payload[1] & 0xFF) << 16; //8 x 2
+        dataLength += (payload[2] & 0xFF) << 8; //8 x 1
+        dataLength += payload[3] & 0xFF; //8 x 0
 
+        let frameCount = (payload[4] & 0xFF) << 24;
+        frameCount += (payload[5] & 0xFF) << 16;
+        frameCount += (payload[6] & 0xFF) << 8;
+        frameCount += payload[7] & 0xFF;
+
+        this._totalConsecutiveFrames = frameCount;
+        this._consecutiveFramesDataLength = dataLength;
     }
-    
+
     /**
-    * @param {SdlPacket} sdlPacket
-    */
-    handleRemainingFrame(sdlPacket){
+     * First frame should be read in before this is called. Once the last
+     * frame is read, callback is used.
+     * @param {SdlPacket} sdlPacket 
+     */
+    _handleConsecutiveFrame(sdlPacket) {
+        this._consecutiveFramesHandledCount++;
+
+        const frameType = sdlPacket.getFrameType();
 
+        //frameSequence nth frame, or 0 for last frame,
+        //8 bits are used to store this info so the count reverts to 1 after 255
+        //This doesn't affect reading the packet but it is notable if frame sequence is relied on.
+        const frameSequence = sdlPacket.getFrameInfo();
+
+        let payload = sdlPacket.getPayload();
+
+        for (let byte of payload) {
+            this._accumulator.push(byte);
+        }
+
+        //This is the last frame, finish things up and use callback.
+        if (frameSequence === 0) {
+            if (this._consecutiveFramesHandledCount !== this._totalConsecutiveFrames) {
+                console.warn(`Total frames expected ${this._totalConsecutiveFrames} does not match total frames received ${this._consecutiveFramesHandledCount}`);
+            }
+
+            if (this._consecutiveFramesDataLength !== this._accumulator.length) {
+                console.warn(`Total data expected ${this._consecutiveFramesDataLength} does not match total data received ${this._accumulator.length}`);
+            }
+
+            let finalData = Uint8Array.from(this._accumulator);
+
+            let dataSize = finalData.length;
+
+            let finishedSdlPacket = new SdlPacket(
+                sdlPacket.getVersion(),
+                sdlPacket.getEncryption(),
+                frameType,
+                sdlPacket.getServiceType(),
+                frameSequence,
+                sdlPacket.getSessionID(),
+                dataSize,
+                sdlPacket.getMessageID(),
+                this._accumulator,
```

Comment:
shouldn't this be `finalData`?

Reply:
The accumulator already has a unit8 array stored and doesn't need to be converted. I removed that extra conversion step, so finalData isn't needed anymore.